packages/server-wallet/src/onchain-service/onchain-service.ts

```diff
@@ -0,0 +1,297 @@
+          return;
+        }
+
+        // Get asset holder information from chain
+        const info = await getAssetHolderInformation(assetHolderAddr, this.provider);
+        this._registerAssetHolderCallbacks(info);
+      })
+    );
+  }
+
+  public async submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse> {
+    if (!this.channelWallet) {
+      throw new OnchainServiceError(OnchainServiceError.reasons.noChannelWallet);
+    }
+
+    // If the channel is not registered, do not send transactions
+    if (!this.store.has(channelId)) {
+      throw new OnchainServiceError(OnchainServiceError.reasons.notRegistered);
+    }
+
+    return this.transactionSubmissionService.submitTransaction(tx, {
+      maxSendAttempts: this.attempts,
+    });
+  }
+
+  private _createDepositEvt(): Evt<DepositedEvent & {block: number}> {
+    // Create the evt instances for the contract
+    const depositEvt = Evt.create<DepositedEvent & {block: number}>();
+
+    // Setup deposit evt so it will emit properly formatted event IFF
+    // - belongs to a registered channel
+    // - is for a subsequent deposit event
+    // And execute channel wallet and storage callbacks (wallet callback
+    // executed IFF a channel wallet has been attached to the service)
+    depositEvt
+      .pipe(e => {
+        return this.store.has(e.destination);
+      })
+      .pipe(e => {
+        const record = this.store.get(e.destination);
+        const prevHoldings = record?.Deposited?.destinationHoldings || BigNumber.from(0);
+        return prevHoldings.lt(e.destinationHoldings);
+      })
+      .attach(e => {
+        // Store updated event record
+        const record = this.store.get(e.destination);
+        const updated: ChannelEventRecord = {
+          ...record,
+          Deposited: {
+            ...e,
+          },
+        };
+        this.store.set(e.destination, {...updated});
+      });
+    return depositEvt;
+  }
+
+  // NOTE: could also do this in the `wait` callback in `submitTransaction`
+  // but would need more information about what event to parse from the
+  // receipt via the API
+  private _registerAssetHolderCallbacks(info: AssetHolderInformation): void {
+    // Create the evt instances for the contract
+    const depositEvt = this._createDepositEvt();
+
+    // Attach the wallet callback
+    depositEvt.attach(e => {
+      // Call the appropriate callback on the wallet
+      this.channelWallet &&
+        this.channelWallet.updateChannelFunding({
+          channelId: e.destination,
+          amount: e.amountDeposited.toString(),
+          token: info.tokenAddress,
+        });
+    });
+
+    // Post to evt on every onchain deposit event
+    const contract = new Contract(
+      info.assetHolderAddress,
+      info.tokenAddress === constants.AddressZero
+        ? ContractArtifacts.EthAssetHolderArtifact.abi
+        : ContractArtifacts.Erc20AssetHolderArtifact.abi,
```

Comment:
I think this is a pretty critical mapping, and we may want to organize it somewhere else in the folder

Reply:
I'm not sure what you mean here, can you explain a bit more?

Reply:
The mapping from `tokenAddress` to asset holder ABI might not be as simple as this. I can think of one example where you might have an asset holder for a token that is using neither `EthAssetHolderArtifact` nor `Erc20AssetHolderArtifact`. I'm not sure where else we might want to put this, but it probably shouldn't be inline.

Reply:
Okay, we can separate it into a function but doing anything more than that is just premature since we only have an `EthAssetHolderArtifact` and an `Erc20AssetHolderArtifact`

Reply:
For TG we have a unique asset holder which is neither of these two